numpy/core/overrides.py

```diff
@@ -44,43 +58,62 @@ def get_overloaded_types_and_args(relevant_args):
+        Iterable of arguments to check for __array_function__ methods.
+    args : tuple
+        Arbitrary positional arguments originally passed into ``public_api``.
+    kwargs : tuple
+        Arbitrary keyword arguments originally passed into ``public_api``.
+
+    Returns
+    -------
+    Result from calling `implementation()` or an `__array_function__`
+    method, as appropriate.
+
+    Raises
+    ------
+    TypeError : if no implementation is found.
+    """
+    # Check for __array_function__ methods.
+    types, overloaded_args = get_overloaded_types_and_args(relevant_args)
+    if not overloaded_args:
+        return implementation(*args, **kwargs)
+
+    # Call overrides
     for overloaded_arg in overloaded_args:
-        # Note that we're only calling __array_function__ on the *first*
-        # occurence of each argument type. This is necessary for reasonable
-        # performance with a possibly long list of overloaded arguments, for
-        # which each __array_function__ implementation might reasonably need to
-        # check all argument types.
-        result = overloaded_arg.__array_function__(func, types, args, kwargs)
+        # Use `public_api` instead of `implemenation` so __array_function__
+        # implementations can do equality/identity comparisons.
+        result = overloaded_arg.__array_function__(
+            public_api, types, args, kwargs)
 
         if result is not NotImplemented:
             return result
 
     raise TypeError('no implementation found for {} on types that implement '
                     '__array_function__: {}'
-                    .format(func, list(map(type, overloaded_args))))
+                    .format(public_api, list(map(type, overloaded_args))))
 
 
 def array_function_dispatch(dispatcher):
-    """Wrap a function for dispatch with the __array_function__ protocol."""
-    def decorator(func):
-        @functools.wraps(func)
-        def new_func(*args, **kwargs):
-            # Collect array-like arguments.
-            relevant_arguments = dispatcher(*args, **kwargs)
-            # Check for __array_function__ methods.
-            types, overloaded_args = get_overloaded_types_and_args(
-                relevant_arguments)
-            # Call overrides, if necessary.
-            if overloaded_args:
-                # new_func is the function exposed in NumPy's public API. We
-                # use it instead of func so __array_function__ implementations
-                # can do equality/identity comparisons.
-                return array_function_override(
-                    overloaded_args, new_func, types, args, kwargs)
-            else:
-                return func(*args, **kwargs)
-
-        return new_func
+    """Decorator for adding dispatch with the __array_function__ protocol."""
+    def decorator(implementation):
+        @functools.wraps(implementation)
+        def public_api(*args, **kwargs):
+            relevant_args = dispatcher(*args, **kwargs)
```

Comment:
I moved the call to `dispatcher(*args, **kwargs)` into `array_function_dispatcher()` to leave things slightly more flexible for `array_function_implementation_or_override()`, e.g., for functions with complex argument parsing such as `np.einsum()`.